### ISSUE TITLE
💡 [Feature]: `Deploy project` Automatically Update the Application after Deployed to App Catalogue

### ISSUE THREAD

**Issue by 6gal6ler6** [pnp/vscode-viva#243]
### 🎯 Aim of the feature

First, I want to thank all people who have been contributing to this amazing tool.

Now, lets get into my improvement idea, for an already amazing feature:

When using the feature `Deploy project`, it would be great if this could also allow the user to update the Application.

I am not 100% sure if this is something that could be implemented.

Still, I think this will benefit developers by adding that as an option for them, so they don’t have to go to the SPO Site Catalogue to update the App every time the package is deployed.

## Feature:

![image](https://github.com/pnp/vscode-viva/assets/20881746/6b55ab91-136f-42a2-8a8a-45bfcf6f42db)

**Comment by Adam-it** [comment]
Thank you for opening up this issue and for the positive feed.
I will try to research this and propose an implementation solution and in which milestone: `v3.X` or `v4.` it may fit

**Comment by 6gal6ler6** [comment]
@Adam-it 

This might help:


I created this small function with `PnP PowerShell` that now I am Using in my App Deployment to update the app.


```PowerShell

function Update-SharePointApp {
    param(
        [Parameter(Mandatory=$true)]
        [string]$SharePointSite,
        [Parameter(Mandatory=$true)]
        [string]$AppName
    )

    try {
        # Connect to SharePoint Online
        Write-Host "Connecting to SharePoint Online site: $SharePointSite"
        Connect-PnPOnline -Url $SharePointSite #ADD CONNECTION METHOD

        # Get the app
        Write-Host "Getting the app: $AppName"
        $App = Get-PnPApp -Scope Site | Where-Object {$_.Title -eq $AppName}

        # Update the app
        Write-Host "Updating the app..."
        Update-PnPApp -Identity $App.Id -Scope Site

        Write-Host "App updated successfully."
    }
    catch {
        Write-Host "An error occurred: $_"
    }
}



Update-SharePointApp -SharePointSite "<MY_COOL_SPO_SITE>";

```

**Comment by nicodecleyre** [comment]
Hi @6gal6ler6, @Adam-it,

Should be able with using `spo app upgrade` in `src\services\CliActions.ts` (deploy action)

**Comment by Adam-it** [comment]
yes indeed what you @nicodecleyre suggest is the way to go. First of all SPFx Toolkit uses CLI for M365 under the hood so we may use the `spo app upgrade` command. 
@6gal6ler6 thank you very much for the additional comment and research you gave. Unfortunately, PnP PS is not an option for SPFx Toolkit as PowerShell may not be used in node.js app directly and we may not also assume user will have PowerShell 7 installed on his local workspace. So we need to be using node.js based approach like CLI for M365.

I am still thinking/prototyping this feature in my head 😜. I apologize it is taking sooo long but I am trying to organize the chaos in my mind to wrap my head around more work. 
the biggest challenge here is that `spo app upgrade` requires to siteUrl to be passed. This by default will extend this action with an additional step one might no always want to do so we should for sure:
1. allow it to be customized turn on/off in extension settings
2. add support for the site Url to be picked rather than type by hand/keyboard.

**Comment by Adam-it** [comment]
@6gal6ler6 so we've picked a bit different approach. We started working on more manage capabilities of apps and currently it is possible to list all apps deployed to any app catalog and perform some basic actions like enable/disable remove etc. 
We are planning to add more and actually the `upgrade` will be one of the actions we are planning to add. You may read more about this feature in this issue #324
Please have a check on that and let me know if it would cover your case, and this kind of solution where you first deploy and then run the upgrade action as a separate action from the context menu next to the app in the app catalog list would work for you as well?

**Comment by Adam-it** [comment]
We added the `upgrade` and now even `install` and `uninstall`  actions which may be done to any SPFx app in any app catalog so I guess the scenario described in this issue is covered. 
We also have a PR opened for a action which will allow you to increase versioning of the SPFx app
